doc/api/async_hooks.md

```diff
@@ -2,7 +2,16 @@
 
 <!--introduced_in=v8.1.0-->
 
-> Stability: 1 - Experimental
+> Stability: 1 - Experimental. Please migrate away from this API, if you can.
+> We do not recommend using the [`createHook`][], [`AsyncHook`][], and
+> [`executionAsyncResource`][] APIs as they have usability issues, safety risks,
+> and performance implications. Async context tracking use cases are better
```

Comment:
Using any tracing API has performance implications, but the overhead of using `createHook()` is on par with other diagnostics APIs. I recommend you remove this.

Reply:
Also I'm curious what safety risks are being referred to here. Please have an example before blaming the API.

Reply:
According to @Qard `createHook` is far more memory intensive than `AsyncLocalStorage`. I’ll leave it to him to cite sources, but I don’t think we need to back up this claim in the docs.

The safety risks are that `createHook`‘s hooks expose Node internals activity.

Reply:
> According to @Qard createHook is far more memory intensive than `AsyncLocalStorage`.

This is so easily verifiably false I'm not sure what to say. As someone who uses the `createHook()` API regularly I can assure you this is not true.

> The safety risks are that `createHook`‘s hooks expose Node internals activity.

As opposed to one of the many other ways there are to get access to Node internals? If we're taking about execution order of the internal API, that's actually very helpful when debugging across versions. It's a feature, not a security risk.

Reply:
That's not a correct characterization of what I said. AsyncLocalStorage is built on top of async_hooks currently, so it can only be as fast as async_hooks can. What I tried to explain in our conversations before is that how users typically _use_ async_hooks is a lot slower than how AsyncLocalStorage uses it. 99% of async_hooks use is a reimplementation of context management and it almost universally uses the destroy hook for cleanup rather than `executionAsyncResource()` which can be _hundreds_ of times slower in promise-heavy applications due to needing to attach garbage collection hooks to every seen promise.

As for the risks, async_hooks exposes internal handles which is a security and stability risk as consumers can access and interact with internals they should not have access to. I've seen many examples of users doing bad things with the API and crashing the process. The typical modelling of context storage _over_ async_hooks is also can very easily become a source of memory leaks if not handled very carefully. The fact that internal handles are _currently_ accessible in a bunch of ways is not a valid argument for us to continue to expose them. They are undocumented and not supported as part of the public versioned API thus they could change or be removed at any time. People should not be interacting with those internals directly _ever_. If there's a use case there, we should make an actual API for it, not have some unwritten rule that users can just do whatever they want with our internals.

Reply:
> AsyncLocalStorage is built on top of async_hooks currently, so it can only be as fast as async_hooks can. What I tried to explain in our conversations before is that how users typically use async_hooks is a lot slower than how AsyncLocalStorage uses it

This explanation is totally fair, but is not the way it comes across in the PR. The current wording makes it sound like using async hooks in any way will have performance implications. Maybe better wording that explains not to roll-your-own implementation because core has made sure AsyncLocalStorage is both safer and more performant. And IMO it would be more convincing for users than just a blanket statement that seems like it's trying to scare me away.

> People should not be interacting with those internals directly ever.

The problem here is that there's a use case for it. At least it has been for me. Observing internal object instances has been of great help when debugging async issues. I agree that users shouldn't be able to alter the internal objects, but I was never able to come up with a solution for that.

Keep in mind that tracking a request object can easily be lost when using AsyncLocalStorage if the request is moved between different request queues for something like batching. This is a real life example where I needed to track every async stack for basically everything and piece together where the request went because AsyncLocalStorage couldn't track it correctly.

Reply:
If you have use cases for visibility into internals we should define APIs specifically for that need and expose that data in an immutable way, security-gate it where relevant, and identify which parts we should gate behind `--expose-internals` and which parts we should consider promoting to public API surface. The current state is very nebulous and as a result has a whole lot of footguns that users run into constantly.

I'm happy to help define and build better APIs, but we need to know the specifics of any use cases that `AsyncLocalStorage` does not solve so that we can satisfy those requirements in a more purpose-built way. I respect that async_hooks was a reasonable approach at the time given the small-core ethos and lack of resources to build out many more specifically-targeted APIs. However we've been pushing toward improving security and usability of the platform in recent years and async_hooks has lagged behind in that regard. It needs to evolve to the next iteration, and likely aspects of its use like AsyncLocalStorage should be standardized, so we should be putting effort into defining what that direction is and pushing forward rather than just defending the status quo of users unhappy their app is crashing or losing context because someone didn't understand a specific edge case of async_hooks behaviour.

I don't expect us to remove async_hooks any time soon, if ever. It is extensively depended on in the ecosystem so removing it would anger a whole lot of users. However, I _do_ expect us to work towards designing more suitable solutions wherever possible so we can move the ecosystem in a direction of greater usability and security. A very small number of async_hooks users actually understand how it works and genuinely _need_ the data it provides. The vast majority are using it (badly) to solve the context propagation problem which, until relatively recently, didn't have a more suitable solution. We now _have_ that solution, and I intend to define solutions for other use cases for async_hooks so we can build new interfaces which are more secure and more usable in their specific scenarios. It's likely the code of async_hooks would still serve these higher-level interfaces, so it's not like it will be going away any time soon. I'm just trying to help guide users to stop doing dangerous things and use tools actually designed for solving their problem.

Reply:
I’m not sure where this thread is going or if anything in the text needs to change. I’m not really interested in getting into fine details of when `createHooks` might be okay, or what types of usage have greater or lesser performance hits, or when it might be permissible to observe internals; the simple message is that we discourage using these APIs, full stop, and reach out to us if you can’t migrate for some reason. The current text says that, and if there’s some way to improve it without getting into minutiae I’m open to suggestions, but making the text less “scary” is not my goal.

Reply:
@trevnorris I'm going to consider this resolved unless I hear from you. Based on this conversation I don't see anything actionable to do.